File: medagent-frontend/src/types/evaluation.ts
export type RunStatus = 'queued' | 'running' | 'completed' | 'failed';
export type GateStatus = 'pending' | 'pass' | 'fail';

export interface EvaluationDataset { id: string; name: string; version: string; description?: string; case_count: number }
export interface DatasetVersion { id:string; version:string; case_count:number; change_summary?:string; created_by?:number; created_at:string }
export interface DatasetHealth { dataset_id:string; version:string; case_count:number; health_score:number; health_score_definition?:string; coverage:number; category_distribution:Record<string,number>; difficulty_distribution:Record<string,number>; duplicate_pairs:number; duplicate_rate:number; drift_js:number|null; discriminative_power:number|null; discriminative_power_status?:string; authoritative_ground_truth_rate:number|null; ground_truth_status?:string }
export interface EvaluationCandidate { id:string; query_text:string; source_request_id?:string; cluster_key?:string; category:string; difficulty:string; status:'pending'|'approved'|'rejected'|'added'; score:number; score_breakdown:Record<string,number>; runtime_metadata:Record<string,unknown>; review_comment?:string; reviewed_by?:number; reviewed_at?:string; target_dataset_id?:string; target_case_id?:number; created_at:string }
export interface HybridModeCost { mode:string; task_completion:number; answer_quality:number; avg_tokens:number; avg_llm_calls:number; avg_tool_calls:number; avg_cost:number; avg_latency_ms:number }
export interface HybridRuntimeMetrics { direct_rate:number; react_rate:number; multi_agent_rate:number; escalation_rate:number; routing_accuracy:number; execution_mode_accuracy:number; under_routing_rate:number; over_routing_rate:number; unnecessary_multi_agent_rate:number; token_savings?:number; cost_savings?:number; latency_savings?:number; quality_vs_cost:HybridModeCost[] }
export interface LiveRetrievalBenchmark { schema_version:string; evaluated_at:string; dataset:string; dataset_version?:string; metric_definition_version?:string; verification_status?:string; status:'healthy'|'degraded'|'missing'; sample_size:number; successful_queries:number; degraded_queries:number; k:number; metrics:{ hit_rate_at_5:number; precision_at_5:number; recall_at_5:number; mrr:number; ndcg_at_5:number; map_at_5?:number }; comparison?:{label:string;before:{metrics:Record<string,number>};after:{metrics:Record<string,number>};delta:Record<string,number>}; routes:Record<string,{available:boolean;reason_code?:string}>; limitations?:string[] }
export interface ReleaseGateResult { status: GateStatus; passed?: boolean; reasons: string[]; config?: Record<string, number>; critical_safety_failure?: boolean }
export interface EvaluationRun {
  id: string; name: string; dataset_id: string; dataset_name?: string; status: RunStatus; mode: 'FAST' | 'FULL' | 'CUSTOM';
  architecture?: 'HYBRID'|'SIMPLE_RAG'|'ALL_MULTI_AGENT'; is_baseline?: boolean;
  agent_version: string; model_name: string; model_version?: string; prompt_version?: string; embedding_model?: string; git_commit?: string;
  started_at?: string; finished_at?: string; created_at: string; duration_ms?: number; total_cases: number; completed_cases: number;
  passed_cases: number; failed_cases: number; critical_failures: number; overall_score?: number; pass_rate: number;
  release_gate_status: GateStatus; release_gate: ReleaseGateResult; error?: string; metrics?: EvaluationMetricMap;
}
export type EvaluationMetricMap = Record<string, number>;
export interface EvaluationCaseResult {
  id: number; run_id: string; case_id: string; case_key: string; name: string; category: string; difficulty: string; tags: string[];
  is_critical: boolean; input: string; expected: EvaluationExpected; thresholds: Record<string, number>; trace_id?: string; status: 'pass' | 'fail' | 'critical';
  passed: boolean; critical: boolean; overall_score: number; scores: EvaluationMetricMap; actual_output?: string; failure_reason?: string;
  judge_reason?: string; details: Record<string, Record<string, unknown>>; latency_ms: number; input_tokens: number; output_tokens: number; estimated_cost: number;
}
export interface EvaluationExpected { facts?: string[]; documents?: string[]; agents?: string[]; tools?: string[]; tool_args?: Record<string, unknown>; trajectory?: string[]; required_content?: string[]; forbidden_content?: string[]; execution_mode?: string|string[]; complexity?: string }
export interface AgentStep { step_number: number; agent_name: string; action: string; input?: unknown; output?: unknown; latency_ms: number; status: string }
export interface EvaluationToolCall { tool_name: string; arguments: Record<string, unknown>; result?: unknown; success: boolean; error?: string; latency_ms: number }
export interface EvaluationRetrieval { query: string; document_id: string; chunk_id: string; rank: number; similarity_score: number; content: string }
export interface EvaluationTrace { id: string; run_id: string; request_id: string; agent_version?: string; model_name?: string; prompt_version?: string; input_text: string; final_output?: string; agent_steps: AgentStep[]; routing_history: string[]; retrievals: EvaluationRetrieval[]; tool_calls: EvaluationToolCall[]; llm_calls: Array<Record<string, unknown>>; input_tokens: number; output_tokens: number; total_tokens: number; estimated_cost: number; latency_ms: number; status: string; error?: string; metadata: Record<string, unknown>; started_at: string; ended_at: string }
export interface EvaluationBaseline { run:EvaluationRun; metrics:EvaluationMetricMap; definition:string; comparison:EvaluationMetricMap }
export interface DashboardData { latest_run: EvaluationRun | null; metrics: EvaluationMetricMap; deltas: EvaluationMetricMap; baseline:EvaluationBaseline|null; trends: Array<{ run_id: string; name: string; date: string; overall_score: number } & EvaluationMetricMap>; recent_runs: EvaluationRun[]; release_gate: ReleaseGateResult | null; hybrid_runtime:HybridRuntimeMetrics; dataset_health:DatasetHealth; pending_candidates:number; live_retrieval:LiveRetrievalBenchmark }
export interface EvaluationComparison { baseline: EvaluationRun; candidate: EvaluationRun; metrics: Array<{ metric: string; baseline: number; candidate: number; delta: number }>; comparison: { summary: Record<string, number>; cases: Array<{ case_key: string; category: string; baseline_score: number; candidate_score: number; delta: number; classification: string; critical: boolean; reason?: string }> }; release_gate: ReleaseGateResult }
export interface PageResult<T> { items: T[]; page: number; page_size: number; total: number }

File: medagent-backend/app/models/evaluation.py
"""Persistent evaluation datasets, executions, metrics and privacy-safe traces."""

from sqlalchemy import BigInteger, Boolean, Column, DateTime, Float, ForeignKey, Integer, JSON, String, Text, UniqueConstraint, func

from app.db.base import MySQLBase

EVAL_PK = BigInteger().with_variant(Integer, "sqlite")


class EvalDataset(MySQLBase):
    __tablename__ = "eval_dataset"

    id = Column(String(64), primary_key=True)
    name = Column(String(255), nullable=False)
    version = Column(String(64), nullable=False)
    description = Column(Text, nullable=True)
    source = Column(String(32), nullable=False, default="database")
    case_count = Column(Integer, nullable=False, default=0)
    created_at = Column(DateTime, server_default=func.now())


class EvalCase(MySQLBase):
    __tablename__ = "eval_case"
    __table_args__ = (UniqueConstraint("dataset_id", "case_key", name="uq_eval_case_dataset_key"),)

    id = Column(EVAL_PK, primary_key=True, autoincrement=True)
    dataset_id = Column(String(64), ForeignKey("eval_dataset.id", ondelete="CASCADE"), nullable=False, index=True)
    case_key = Column(String(128), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    category = Column(String(64), nullable=False, index=True)
    difficulty = Column(String(16), nullable=False, default="medium")
    tags = Column(JSON, nullable=False)
    is_critical = Column(Boolean, nullable=False, default=False, index=True)
    input_text = Column(Text, nullable=False)
    expected_json = Column(JSON, nullable=False)
    thresholds_json = Column(JSON, nullable=False)
    source_json = Column(JSON, nullable=False)
    created_at = Column(DateTime, server_default=func.now())


class EvalRun(MySQLBase):
    __tablename__ = "eval_run"

    id = Column(String(64), primary_key=True)
    name = Column(String(255), nullable=False)
    dataset_id = Column(String(64), ForeignKey("eval_dataset.id"), nullable=False, index=True)
    status = Column(String(24), nullable=False, default="queued", index=True)
    mode = Column(String(16), nullable=False)
    agent_version = Column(String(64), nullable=False)
    model_name = Column(String(128), nullable=False)
    model_version = Column(String(64), nullable=True)
    prompt_version = Column(String(64), nullable=True)
    embedding_model = Column(String(128), nullable=True)
    embedding_version = Column(String(64), nullable=True)
    git_commit = Column(String(64), nullable=True)
    retrieval_config = Column(JSON, nullable=False)
    judge_model = Column(String(128), nullable=True)
    judge_prompt_version = Column(String(64), nullable=True)
    random_seed = Column(Integer, nullable=False, default=0)
    config_json = Column(JSON, nullable=False)
    started_at = Column(DateTime, nullable=True)
    finished_at = Column(DateTime, nullable=True)
    duration_ms = Column(Float, nullable=True)
    total_cases = Column(Integer, nullable=False, default=0)
    completed_cases = Column(Integer, nullable=False, default=0)
    passed_cases = Column(Integer, nullable=False, default=0)
    failed_cases = Column(Integer, nullable=False, default=0)
    critical_failures = Column(Integer, nullable=False, default=0)
    overall_score = Column(Float, nullable=True)
    release_gate_status = Column(String(16), nullable=False, default="pending")
    release_gate_json = Column(JSON, nullable=False)
    error = Column(Text, nullable=True)
    created_at = Column(DateTime, server_default=func.now(), index=True)


class EvalCaseResult(MySQLBase):
    __tablename__ = "eval_case_result"
    __table_args__ = (UniqueConstraint("run_id", "case_id", name="uq_eval_case_result_run_case"),)

    id = Column(EVAL_PK, primary_key=True, autoincrement=True)
    run_id = Column(String(64), ForeignKey("eval_run.id", ondelete="CASCADE"), nullable=False, index=True)
    case_id = Column(BigInteger, ForeignKey("eval_case.id"), nullable=False, index=True)
    case_key = Column(String(128), nullable=False, index=True)
    trace_id = Column(String(64), nullable=True, index=True)
    status = Column(String(24), nullable=False, index=True)
    passed = Column(Boolean, nullable=False)
    critical = Column(Boolean, nullable=False, default=False, index=True)
    overall_score = Column(Float, nullable=False)
    scores_json = Column(JSON, nullable=False)
    actual_output = Column(Text, nullable=True)
    failure_reason = Column(Text, nullable=True)
    judge_reason = Column(Text, nullable=True)
    details_json = Column(JSON, nullable=False)
    latency_ms = Column(Float, nullable=False, default=0)
    input_tokens = Column(Integer, nullable=False, default=0)
    output_tokens = Column(Integer, nullable=False, default=0)
    estimated_cost = Column(Float, nullable=False, default=0)
    created_at = Column(DateTime, server_default=func.now())


class EvalMetricResult(MySQLBase):
    __tablename__ = "eval_metric_result"
    __table_args__ = (UniqueConstraint("run_id", "metric_name", name="uq_eval_metric_run_name"),)

    id = Column(EVAL_PK, primary_key=True, autoincrement=True)
    run_id = Column(String(64), ForeignKey("eval_run.id", ondelete="CASCADE"), nullable=False, index=True)
    evaluator_name = Column(String(128), nullable=False)
    metric_name = Column(String(128), nullable=False, index=True)
    score = Column(Float, nullable=False)
    passed = Column(Boolean, nullable=False)
    details_json = Column(JSON, nullable=False)
    created_at = Column(DateTime, server_default=func.now())


class EvaluationTrace(MySQLBase):
    __tablename__ = "agent_trace"

    id = Column(String(64), primary_key=True)
    run_id = Column(String(64), ForeignKey("eval_run.id", ondelete="CASCADE"), nullable=False, index=True)
    case_result_id = Column(BigInteger, ForeignKey("eval_case_result.id", ondelete="CASCADE"), nullable=True, index=True)
    request_id = Column(String(64), nullable=False, index=True)
    session_id = Column(String(64), nullable=True)
    user_id_hash = Column(String(64), nullable=True)
    agent_version = Column(String(64), nullable=True)
    model_name = Column(String(128), nullable=True)
    prompt_version = Column(String(64), nullable=True)
    input_text = Column(Text, nullable=False)
    final_output = Column(Text, nullable=True)
    agent_steps = Column(JSON, nullable=False)
    routing_history = Column(JSON, nullable=False)
    retrievals = Column(JSON, nullable=False)
    tool_calls = Column(JSON, nullable=False)
    llm_calls = Column(JSON, nullable=False)
    input_tokens = Column(Integer, nullable=False, default=0)
    output_tokens = Column(Integer, nullable=False, default=0)
    total_tokens = Column(Integer, nullable=False, default=0)
    estimated_cost = Column(Float, nullable=False, default=0)
    latency_ms = Column(Float, nullable=False, default=0)
    status = Column(String(24), nullable=False)
    error = Column(Text, nullable=True)
    metadata_json = Column(JSON, nullable=False)
    started_at = Column(DateTime, nullable=False)
    ended_at = Column(DateTime, nullable=False)
    created_at = Column(DateTime, server_default=func.now())


class EvalDatasetVersion(MySQLBase):
    __tablename__ = "eval_dataset_version"
    __table_args__ = (UniqueConstraint("dataset_id", "version", name="uq_eval_dataset_version"),)

    id = Column(String(64), primary_key=True)
    dataset_id = Column(String(64), ForeignKey("eval_dataset.id", ondelete="CASCADE"), nullable=False, index=True)
    version = Column(String(64), nullable=False)
    case_count = Column(Integer, nullable=False, default=0)
    change_summary = Column(Text, nullable=True)
    snapshot_json = Column(JSON, nullable=False)
    created_by = Column(BigInteger, nullable=True)
    created_at = Column(DateTime, server_default=func.now())


class EvaluationCandidate(MySQLBase):
    __tablename__ = "evaluation_candidate"

    id = Column(String(64), primary_key=True)
    normalized_hash = Column(String(64), nullable=False, index=True)
    query_text = Column(Text, nullable=False)
    source_request_id = Column(String(64), nullable=True, index=True)
    cluster_key = Column(String(64), nullable=True, index=True)
    category = Column(String(64), nullable=False, default="production_failure", index=True)
    difficulty = Column(String(16), nullable=False, default="medium")
    status = Column(String(24), nullable=False, default="pending", index=True)
    score = Column(Float, nullable=False, default=0, index=True)
    score_breakdown = Column(JSON, nullable=False)
    runtime_metadata = Column(JSON, nullable=False)
    review_comment = Column(Text, nullable=True)
    reviewed_by = Column(BigInteger, nullable=True)
    reviewed_at = Column(DateTime, nullable=True)
    target_dataset_id = Column(String(64), ForeignKey("eval_dataset.id"), nullable=True)
    target_case_id = Column(BigInteger, ForeignKey("eval_case.id"), nullable=True)
    created_at = Column(DateTime, server_default=func.now(), index=True)


class DatasetChangeLog(MySQLBase):
    __tablename__ = "dataset_change_log"

    id = Column(EVAL_PK, primary_key=True, autoincrement=True)
    dataset_id = Column(String(64), ForeignKey("eval_dataset.id", ondelete="CASCADE"), nullable=False, index=True)
    version = Column(String(64), nullable=False)
    action = Column(String(32), nullable=False)
    case_key = Column(String(128), nullable=True)
    candidate_id = Column(String(64), nullable=True)
    actor_id = Column(BigInteger, nullable=True)
    details_json = Column(JSON, nullable=False)
    created_at = Column(DateTime, server_default=func.now())
